### ISSUE TITLE
Making PittAPI into a proper package

### ISSUE THREAD

**Issue by azharichenko** [Pitt-CSC/PittAPI#23]
## Proposal:
Essentially the change I proposing is moving all of PittAPI code into a folder and modularizing the project. This includes a big move away from OOP programming since there isn't much use of it in this library (this will get explained towards the end). So I'm going to show how the API will look after the change, how it will get structured, and why to get rid of the OOP usage.

## API Usage Change:
So let's look at an example from of the current usage from the README.md of how to use the course API...

```python
from pittAPI import CourseAPI

### Courses
course = CourseAPI()
# Will return a list of dictionaries containing courses in subject
big_dict = course.get_courses(term="2161", subject="CS")
# Will return a list of dictionaries containing courses that fulfill req
big_dict = course.get_courses_by_req(term="2161", req="Q")
# Will return a string, which is the description of the course
description = course.get_class_description(class_number="10163", term="2161")
```

and here is what it will look like after the change.

```python
from PittAPI import course

# Will return a list of dictionaries containing courses in subject
big_dict = course.get_courses(term="2161", subject="CS")
# Will return a list of dictionaries containing courses that fulfill req
big_dict = course.get_courses_by_req(term="2161", req="Q")
# Will return a string, which is the description of the course
description = course.get_class_description(class_number="10163", term="2161")

```

### Details to Notice

- Notice how the naming is changed especially with `CourseAPI` turning into  `course`.  Thought behind this is that the `PittAPI` contains different modules that have to access various data from Pitt. Saying `API` repeatedly is a bit redundant since it's all coming from the `PittAPI`, instead it's shorten down to just `course` which is imported from `PittAPI`.
- Next, notice how you no longer have to initialize the `course` API. Instead, you can use it's function directly, making the API even more friendly. This is a change to all the APIs where initialization is no longer required. (I will go into more in-depth why in a later explanation)

## Package Structure:
The way the repository is structured is there is a `pittAPI.py` that has the whole library. First, change is to move this into a folder a.k.a. a _package_. This will come in handy later if this package gets put on the _Python Package Index_ since, in the setup of submitting, you should already have the code all in a folder (many python packages contain their code into packages like this).

#### Before
```
__init__.py
pittAPI.py
```

#### After
```
PittAPI/
  - __init__.py
  - pittAPI.py
```

Very simple change, but in the current state shown above it makes the usage worse but there is one more change I would like to make.

## Moving away from OOP (Making them into modules):
So as I was looking throughout this library I couldn't help but notice the overuses of classes when none are needed. This might seem weird but here me out.

Let's look at that `CourseAPI` class. One issue is why is `__init__` declared since it's not used? Ok that's a bit nitpicky.
```python
class CourseAPI:
    def __init__(self):
        pass
```

But there is a much bigger observation to make with this class. Besides the functions annotated with `@staticmethod`. Such as...

```python
@staticmethod
def _retrieve_from_url(url):
    page = s.get(url)
    soup = BeautifulSoup(page.text, 'html.parser')
    courses = soup.findAll("tr", {"class": "odd"})
    courses_even = soup.findAll("tr", {"class": "even"})
    courses.extend(courses_even)
    return courses
```

Aren't all the functions static? Take a look at this function below notice how there isn't really any usage of `self` (besides using it to call a static method) and that this function is very self-contained. What's the point of forcing all these static functions into a class, in python, there is a better way of doing this.

```python
def get_courses(self, term, subject):
      subject = subject.upper()

      url = 'http://www.courses.as.pitt.edu/results-subja.asp?TERM={}&SUBJ={}'.format(term, subject)
      courses = self._retrieve_from_url(url)

      course_details = []

      for course in courses:
          details = [course_detail.string.replace('&nbsp;', '').strip()
                     for course_detail in course
                     if course_detail.string is not None]

          # Only append details if the list is not empty
          # If the subject code is incorrect, details will be NoneType
          if details:
              course_details.append(
                  {
                      'subject': details[1] if details[1] else "Not Decided",
                      'catalog_number': details[3] if details[3] else "Not Decided",
                      'term': details[5].replace('\r\n\t', '') if details[5] else "Not Decided",
                      'class_number': course.find('a').contents[0] if course.find('a').contents[0] else "Not Decided",
                      'title': details[8] if details[8] else "Not Decided",
                      'instructor': details[10] if details[10] else "Not Decided",
                      'credits': details[12] if details[12] else "Not Decided"
                  }
              )

      if not course_details:
          raise InvalidParameterException("The TERM or SUBJECT is invalid")

      return course_details
```

This better way is to turn these classes into modules. Essentially all that is being done is dropping the `class` and moving the all the functions into their own python file. So the structure will go from this ...

#### Before
```
PittAPI/
  - __init__.py
  - pittAPI.py
```

to this with each API have its own python file.

#### After
```
PittAPI/
  - __init__.py
  - course.py
  - laundry.py
  - people.py
  - lab.py
  - dining.py
```

In an instant, this removes all the OOP and makes the project more manageable, with each API having its own module to contain all its functions and any other specifics that apply to that API. This makes calling the API even simpler since there is no longer the need to initialize a class. It also makes importing the API very simple as all that needs to be done is to import the module.

---

I have already done this and tested it a bit ([take a look at my fork of this project](https://github.com/azharichenko/PittAPI)), but I would love to hear what your opinion of this idea is. If it all sounds good I would be more than glad to do a pull request of all these changes.

---
#### Who am I?
This is all who wonder why someone outside of Pitt is contributing to this project. Hello, I'm Alex Zharichenko and I'm a senior in high school who is currently in his 4th year of computer science (in my school this is basically independent study). I just recently after more than 6 months of waiting (applied on 7/21/16) finally got accepted into Pitt. I'm already fully committed to going to Pitt with my major being Computer Science. Hence the sudden interest into this repository.

**Comment by timparenti** [comment]
+1 and congrats on your acceptance!

**Comment by RitwikGupta** [comment]
@azharichenko The reason for the modularity is that I had plans to store a session in the objects. You might notice that in some cases, such as the Laundry API, we need to get a cookie. Right now I get the same cookie for every call, wasting an expensive I/O call. The plan was to just store the cookie for a period of time in an object, so I kept every class uniform.

However, I like this change. We can look into storing a session for the LaundryAPI later because the overall benefit we get from this is greater than the potential for what I had in mind. Please go ahead with the PR, and congrats to your acceptance to Pitt 😄

**Comment by azharichenko** [comment]
@RitwikGupta Yeah I think we can work out a way with storing a session for LaundryAPI, I'll probably look into that later. I'm right now putting my attention on the CourseAPI since issue #22 brought up some interesting cases that need to be dealt with, plus I think I made more efficient ways of getting the courses but I'm still testing.

Anyways here comes the PR. :tada:

**Comment by RitwikGupta** [comment]
Closed with 759390f12c4e4725dd815bdb0dd1a28534c7b6e8
